pangeo/pbs.py

```diff
@@ -0,0 +1,208 @@
+from contextlib import contextmanager
+import logging
+import os
+import socket
+import subprocess
+import sys
+
+from distributed import LocalCluster
+from distributed.utils import tmpfile, get_ip_interface
+
+template = """
+#!/bin/bash
+#PBS -N %(name)s
+#PBS -q %(queue)s
+#PBS -A %(project)s
+#PBS -l %(resouce_spec)s
+#PBS -l walltime=%(walltime)s
+#PBS -j oe
+#PBS -m abe
+
+%(base_path)s/dask-worker %(scheduler)s \
+    --nprocs %(workers_per_node)d \
+    --nthreads %(threads_per_worker)d \
+    --memory-limit %(memory)s \
+     %(extra)s
+"""
+
+
+logger = logging.getLogger(__name__)
+
+dirname = os.path.dirname(sys.executable)
+
+
+class PBSCluster(object):
+    """ Launch Dask on a PBS cluster
+
+    Examples
+    --------
+    >>> from pangeo import PBSCluster
+    >>> cluster = PBSCluster(project='...')
+    >>> cluster.start_workers(10)  # this may take a few seconds to launch
+
+    >>> from dask.distributed import Client
+    >>> client = Client(cluster)
+
+    This also works with adaptive clusters.  This automatically launches and
+    kill workers based on load.
+
+    >>> from distributed.deploy import Adaptive
+    >>> adapt = Adaptive(cluster.cluster.scheduler, cluster)
+    """
+    def __init__(self,
+                 name='dask',
+                 queue='regular',
+                 project=None,
+                 resouce_spec='select=1:ncpus=36:mem=109GB',
+                 workers_per_node=9,
+                 threads_per_worker=4,
+                 memory='7e9',
+                 walltime='00:30:00',
+                 interface=None,
+                 extra='',
+                 **kwargs):
+        """ Initialize a PBS Cluster
+
+        Parameters
+        ----------
+        name : str
+            Name of worker jobs. Passed to `$PBS -N` option.
+        queue : str
+            Destination queue for each worker job. Passed to `#PBS -q` option.
+        project : str
+            Accounting string associated with each worker job. Passed to
+            `#PBS -A` option.
+        resource_spec : str
+            Request resources and specify job placement. Passed to `#PBS -l`
+            option.
```

Comment:
How should we reconcile this argument and the `memory` and `threads_per_worker` argument?

Also, I'm curious, what was the reason for this change?  I would not be surprised to learn that the previous solution was restrictive in some way.

Also, why the multiple processes per job change?  Is this nicer than having many small jobs?  Would array jobs be more appropriate?  From a cleanup perspective it's nice for the `PBSCluster` to have a handle on each individual worker separately.

Reply:
Short term I've reverted this and gone back to only supporting `--nthreads`.  It is difficult to use `--nprocs`, especially if we want to do adaptive deployments.

Reply:
Yes, motivation for this change was:

1) The resource specification should be flexible enough to move between different machines that use PBS. While `#PBS -l select=1:ncpus=$ncpus:mem=$mem` works on cheyenne, this is not going to work everywhere. I figured it is easy enough to allow users to specify this line.
2) From the [Cheyenne docs](https://www2.cisl.ucar.edu/resources/computational-systems/cheyenne/quick-start-cheyenne):
    > Most of the Cheyenne batch queues are for exclusive use, and jobs are charged for all 36 cores on each node that is used. Jobs in the shared use "share" queue are charged only for the cores used.

    This is why I had setup the PBSCluster to use multiple workers per job. Alternatively, we could a) try using the shared queue, b) accept paying for cores we're not using, or c) use 36 threads per worker (good for cpu bound computations but not IO bound tasks).

Reply:
I suggest avoid option a) at least by default, since sharing node with others can cause a lot of troubles (at least at the moment). Mostly, but not only, because users have workarounds that allow them to use more memory, CPU and network than their share, overloading the node (we plan to deploy a more aggressive watchdog, angrily biting whoever spills over their share, but it's not in place and won't be for a while). Option c) seems best. If you are IO bound, oh well. How many IO bound jobs are a user going to run? For how long total compared to their own total allocation? I suspect for most folks it'd be irrelevant cost. And for the few people it isn't, they could use the non-default a) option and deal with the issues.

Reply:
I also like option (c) from a Dask perspective, however I'm also sensitive to IO bound concerns.  I suspect that our computations are more I/O bound than most that are run on Cheyenne.

@jhamman are you mainly talking about disk I/O or inter-worker network communication?  If disk I/O then do we know what the limiting factor is here?  Is it compression libraries that don't release the GIL?

Reply:
@mrocklin - yes, disk I/O. I suspect it will be difficult to clearly define the limiting factor on disk I/O. It will greatly depend on both how the data is stored and the chunk configuration. One thing we know for sure is that more processes certainly helps.